content/en/blog/_posts/2024-11-11-dynamic-kubernetes-api-server-for-cozystack/index.md

```diff
@@ -0,0 +1,454 @@
+
+{{% alert title="Note" color="info" %}}
+Subresources are additional actions or operations you can perform on primary resources
+(like Pods, Deployments, Services) via the Kubernetes API. They provide interfaces to manage
+specific aspects of resources without affecting the entire object.
+{{% /alert %}}
+
+A simple example is `/status`, which is traditionally exposed as a separate subresource that's
+stored and handled independently from the main resource. The `status` field isn't meant to be changed
+by regular users, so if you want to make any changes to it, your controller has to use a separate
+procedure specifying the subresource name.
+
+But beyond `/status`, Pods in Kubernetes also have subresources like `/exec`, `/portforward`, and
+`/log`. Interestingly, instead of the usual declarative resources in Kubernetes, these represent
+endpoints for imperative operations like viewing logs, proxying connections, executing commands in
+a running container, and so on.
+
+To perform such imperative commands, you implement an extension api-server. Here are some well-known implementations:
+
+- **KubeVirt**: An add-on for Kubernetes that extends its API capabilities to run traditional virtual machines.
+  The extension api-server created as part of KubeVirt handles subresources
+  like `/restart`, `/console`, and `/vnc` for virtual machines.
+- **Knative**: A Kubernetes add-on that extends its capabilities for serverless computing,
+  implementing the `/scale` subresource to set up autoscaling for its resource types.
+
+{{% alert title="Note" color="info" %}}
+By the way, even though subresource logic in Kubernetes can be *imperative*, you can manage access
+to them *declaratively* using Kubernetes standard RBAC model.
+
+For example this way you can control access to the `/log` and `/exec` subresources of the Pod kind:
+```yaml
+kind: Role
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  namespace: default
+  name: pod-and-pod-logs-reader
+rules:
+- apiGroups: [""]
+  resources: ["pods", "pods/log"]
+  verbs: ["get", "list"]
+- apiGroups: [""]
+  resources: ["pods/exec"]
+  verbs: ["create"]
+```
+{{% /alert %}}
+
+### You're not tied to use etcd
+
+Usually, the Kubernetes api-server uses etcd for its backend.
+However, implementing your own api-server doesn't lock you into using only etcd.
+If it doesn't make sense to store your server's state in etcd, you can store information in any
+other system and generate responses on the fly. Here are a few cases to illustrate:
+
+* [Metrics-server](https://github.com/kubernetes-sigs/metrics-server) is a standard extension for Kubernetes
+  which allows you to view real-time metrics of your nodes and pods. It defines alternative Pod and Node
+  kinds in its own metrics.k8s.io API. Requests to these resources are translated into metrics
+  directly from cAdvisor. So when you run `kubectl top node` or `kubectl top pod`, metrics-server fetches
```

Comment:
AIUI, they come indirectly from the kubelet.

Reply:
cAdvisor is integrated into Kubelet

> To generate metrics for pods, Kubernetes employs an agent called kubelet. Kubelet runs on each node and includes a sub-component known as cAdvisor (Container Advisor). cAdvisor is responsible for retrieving metrics from pods and nodes and exposing them through the kubelet API. It collects data such as CPU usage, memory consumption, network statistics, and more.

[source](https://medium.com/@tech_18484/demystifying-kubernetes-metrics-server-monitoring-resource-utilization-in-your-cluster-b7aeaf1b513c)

Reply:
You're not wrong, but it might not be the whole story.
I'm quibbling _directly_, especially as the kubelet can opt to bypass cAdvisor; see https://kubernetes.io/docs/reference/instrumentation/node-metrics/

Reply:
Okay I swaped kubelet and cAdvisor, I think now it is more accurate :-)